sw/device/lib/dif/dif_aes.c

```diff
@@ -222,10 +231,32 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
     return kDifError;
   }
 
-  for (int i = 0; i < AES_DATA_OUT_MULTIREG_COUNT; ++i) {
-    ptrdiff_t offset = AES_DATA_OUT_0_REG_OFFSET + (i * sizeof(uint32_t));
+  aes_read_multireg(aes, data->data, AES_DATA_OUT_MULTIREG_COUNT,
+                   AES_DATA_OUT_0_REG_OFFSET);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_process_data(const dif_aes_t *aes,
+                                  const dif_aes_data_t *plain_text,
+                                  dif_aes_data_t *cipher_text,
+                                  size_t block_amount) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (!aes_input_ready(aes)) {
+    return kDifUnavailable;
+  }
+
+  for (int i = 0; i < block_amount; i++){
+    aes_set_multireg(aes, plain_text[i].data, AES_DATA_IN_MULTIREG_COUNT,
+                    AES_DATA_IN_0_REG_OFFSET);
```

Comment:
This first block should be written outside of the loop just after checking `aes_input_ready`. Then you need another check on `aes_input_ready` before entering the loop.

If you do this, the operation of the AES hardware will be overlapped with processor operation. This is good for performance and side channel hardening.

Reply:
Some more details on how and why can be found in the specification: https://docs.opentitan.org/hw/ip/aes/doc/#block-operation